notebooks/03-query/07-functions-datetime.ipynb

```diff
@@ -0,0 +1,944 @@
+    "  EXTRACT(HOUR FROM MILLIS_TO_TIMESTAMP(\"ended_at\")) AS \"end-hour\",\n",
+    "  COUNT(*) AS \"sessions\"\n",
+    "FROM \"example-koalas-fndatetime\"\n",
+    "GROUP BY 1, 2\n",
+    "'''\n",
+    "\n",
+    "df = pd.DataFrame(sql_client.sql(sql))\n",
+    "df_group=df.groupby(['start-hour','end-hour']).sum().unstack()\n",
+    "df_group.plot.bar(stacked=\"true\", legend=None)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c1ca1ec0-009e-4251-a5c8-27e21d466595",
+   "metadata": {},
+   "source": [
+    "Run the next cell to see how `TIME_EXTRACT` has been used instead of `EXTRACT` to allow a timezone to be specified in the calculation."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "53cb65c0-b650-4fc2-bd3e-00ae9e25bee7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT\n",
+    "  TIME_EXTRACT(__time, 'HOUR') AS \"start-hour\",\n",
+    "  TIME_EXTRACT(__time, 'HOUR', 'America/Los_Angeles') AS \"start-hour-LA\",\n",
+    "  TIME_EXTRACT(__time, 'HOUR', 'India/Kolkota') AS \"start-hour-Ko\",\n",
+    "  COUNT(*) AS \"sessions\"\n",
+    "FROM \"example-koalas-fndatetime\"\n",
+    "WHERE TIME_IN_INTERVAL(__time, '2019-08-25T03/PT2H')\n",
+    "GROUP BY 1,2,3\n",
+    "'''\n",
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f7f4a408-a994-4761-8954-d17dd26ca125",
+   "metadata": {},
+   "source": [
+    "## Rounding timestamps\n",
+    "\n",
+    "There are a number of SQL functions available:\n",
+    "\n",
+    "* Rounding down: `TIME_FLOOR`, `FLOOR`, and `DATE_TRUNC` ([translated](https://druid.apache.org/docs/latest/querying/math-expr#time-functions) to the `timestamp_floor` native JSON function)\n",
+    "* Rounding up: `TIME_CEIL` and `CEIL` (translated to the `timestamp_ceil` native JSON function).\n",
+    "\n",
+    "Run the cell below to see how these functions effect timestamps."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7945a107-d03d-4088-bd3a-07166f11473b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT DISTINCT  \n",
+    "  TIME_FLOOR(__time, 'PT1H') AS \"session_down_hour_1\",\n",
+    "  FLOOR(__time TO HOUR) AS \"session_down_hour_2\",\n",
+    "  DATE_TRUNC('HOUR', __time) AS \"session_down_hour_3\",\n",
+    "  TIME_CEIL(__time, 'PT1H') AS \"session_up_hour_1\",\n",
+    "  CEIL(__time TO HOUR) AS \"session_up_hour_2\"\n",
+    "FROM \"example-koalas-fndatetime\"\n",
+    "WHERE TIME_IN_INTERVAL(__time, '2019-08-25T03/PT2H')\n",
+    "'''\n",
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ed118731-3525-435a-8e6c-003150187a32",
+   "metadata": {},
+   "source": [
+    "Using the more flexible `TIME_FLOOR` and `TIME_CEIL` functions, a parameter has been supplied in the SQL statement below to shift the calculation of the floor / ceiling operation."
```

Comment:
I think this needs more explanation.
The idea is that you are still creating consistent period of time throughout time like 6 hour periods or 12 hour periods or 30 minute periods or whatever, but you while normally you would start at 1970-01-01 00:00:00, by changing the starting point you shift all the floor and ceiling or the calculated periods. Take 6h period for example, 

if it starts at 1970-01-01 00:00:00 then you get:
floor = 1970-01-01 00:00:00 ceiling = 1970-01-01 06:00:00
floor = 1970-01-01 06:00:00 ceiling = 1970-01-01 12:00:00
... so if I look for floor and ceiling of 2023-04-13 17:32:00, I get 2023-04-13 12:00:00 and 2023-04-13 18:00:00 respectively.

If it starts at 1970-01-01 05:00:00 then the you get:
floor = 1970-01-01 05:00:00 ceiling = 1970-01-01 11:00:00
floor = 1970-01-01 11:00:00 ceiling = 1970-01-01 17:00:00 
... so if I look for floor and ceiling of 2023-04-13 17:32:00, I get 2023-04-13 17:00:00 and 2023-04-13 23:00:00 respectively.

Reply:
Added a commentary cell to prompt them to work out how it was calculated, as well as TIME_FLOOR and TIME_CEIL without the additional parameter.